tools/pub_get_offline.py

```diff
@@ -139,7 +154,8 @@ def main():
   for package in ALL_PACKAGES:
     if fetch_package(pubcmd, package) != 0:
       return 1
-    pub_count = pub_count + check_package(package)
+    if not package_uses_workspace_resolution(package):
```

Comment:
Hmm... I'm actually not sure. I _assume_ no, but let me give it a shot.

Reply:
Short answer: Yes, it's possible. It ends up adding it to the _root_ `.dart_tool/package_config.json`.

Long answer: Our script to protect against pub dependencies _still_ works, as it catches the root package config:

```diff
diff --git a/tools/engine_tool/lib/main.dart b/tools/engine_tool/lib/main.dart
index e524e07ad0..d2864ee1db 100644
--- a/tools/engine_tool/lib/main.dart
+++ b/tools/engine_tool/lib/main.dart
@@ -5,6 +5,7 @@
 import 'dart:ffi' as ffi show Abi;
 import 'dart:io' as io show Directory, Platform, exitCode, stderr;
 
+import 'package:archive/archive.dart';
 import 'package:engine_build_configs/engine_build_configs.dart';
 import 'package:engine_repo_tools/engine_repo_tools.dart';
 import 'package:path/path.dart' as p;
@@ -17,6 +18,7 @@ import 'src/logger.dart';
 import 'src/phone_home.dart';
 
 void main(List<String> args) async {
+  print(getAdler32);
   if (phoneHome(args)) {
     return;
   }
diff --git a/tools/engine_tool/pubspec.yaml b/tools/engine_tool/pubspec.yaml
index 1b71b9b688..83759735d5 100644
--- a/tools/engine_tool/pubspec.yaml
+++ b/tools/engine_tool/pubspec.yaml
@@ -13,6 +13,7 @@ environment:
 resolution: workspace
 
 dependencies:
+  archive: any
   args: any
   engine_build_configs: any
   engine_repo_tools: any
```

... and then:

```sh
matanl@matanl-macbookpro4 engine_tool % gclient sync
Syncing projects: 100% (143/143), done.
Running hooks:  46% ( 7/15) pub get --offline   
________ running 'python3 src/flutter/tools/pub_get_offline.py' in '/Users/matanl/Developer/engine'
Error: package "archive" was fetched from pub
Error: package "crypto" was fetched from pub
Error: package "typed_data" was fetched from pub
Error: 3 packages were fetched from pub for /Users/matanl/Developer/engine/src/flutter
Please fix the pubspec.yaml for /Users/matanl/Developer/engine/src/flutter so that all dependencies are path dependencies
Error: Command 'python3 src/flutter/tools/pub_get_offline.py' returned non-zero exit status 1 in /Users/matanl/Developer/engine
Error: package "archive" was fetched from pub
Error: package "crypto" was fetched from pub
Error: package "typed_data" was fetched from pub
Error: 3 packages were fetched from pub for /Users/matanl/Developer/engine/src/flutter
Please fix the pubspec.yaml for /Users/matanl/Developer/engine/src/flutter so that all dependencies are path dependencies
```

I do think the UX of allowing package-specific dependencies and silently adding them to the root pubspec is confusing, but not a blocker for this effort. I'll file an issue on dart-lang/pub and see what the team thinks in terms of UX.

Reply:
Done in https://github.com/dart-lang/pub/issues/4328.

Reply:
Thanks for checking!